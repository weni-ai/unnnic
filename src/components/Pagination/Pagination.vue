<template>
  <div class="pagination">
    <unnnic-button
      type="terciary"
      size="small"
      icon-center="arrow-left-1-1"
      :disabled="!canPrevious || disabled"
      @click="previousPage"
    />

    <unnnic-button
      v-for="(page, index) in pages"
      :key="index"
      :type="page === value ? 'secondary' : 'terciary'"
      size="small"
      :text="['left-hidden', 'right-hidden'].includes(page) ? '...' : String(page)"
      @click="selectPage(page)"
      :disabled="disabled"
    />

    <unnnic-button
      type="terciary"
      size="small"
      icon-center="arrow-right-1-1"
      :disabled="!canNext || disabled"
      @click="nextPage"
    />
  </div>
</template>

<script>
import unnnicButton from '../Button/Button.vue';

export default {
  components: {
    unnnicButton,
  },

  props: {
    value: {
      type: Number,
      default: 1,
    },

    max: {
      type: Number,
      default: 1,
    },

    show: {
      type: Number,
      default: 5,
    },

    disabled: {
      type: Boolean,
      default: false,
    },
  },

  data() {
    return {
      reference: 3,
    };
  },

  computed: {
    pages() {
      const pages = [];

      for (let i = 1; i <= this.max; i += 1) {
        pages.push(i);
      }

      const plus = this.show - 4;
      const maxToRemove = this.max - plus - 3;

      if (this.reference > 3) {
        const toRemove = (this.reference - 2);
        const maxRemove = toRemove > maxToRemove ? maxToRemove : toRemove;
        pages.splice(1, maxRemove, 'left-hidden');
      }

      if (this.reference < this.max - plus - 1) {
        const toRemove = (this.max - plus - this.reference);
        const maxRemove = toRemove > maxToRemove ? maxToRemove : toRemove;
        pages.splice(-(maxRemove + 1), maxRemove, 'right-hidden');
      }

      return pages;
    },

    canPrevious() {
      return this.value > 1;
    },

    canNext() {
      return this.value < this.max;
    },
  },

  methods: {
    selectPage(page) {
      if (page === 'left-hidden') {
        const pageNumber = this.pages[this.pages.indexOf('left-hidden') + 1] - 1;

        this.$emit('input', pageNumber);
        this.setReference(pageNumber);
      } else if (page === 'right-hidden') {
        const pageNumber = this.pages[this.pages.indexOf('right-hidden') - 1] + 1;

        this.$emit('input', pageNumber);

        if (this.pages.includes('left-hidden')) {
          this.setReference(this.pages[this.pages.indexOf('left-hidden') + 1] + 1);
        } else {
          this.setReference(4);
        }
      } else {
        this.$emit('input', page);
      }
    },

    previousPage() {
      if (this.canPrevious) {
        const pageNumber = this.value - 1;

        this.selectPage(pageNumber);

        if (!this.pages.includes(pageNumber)) {
          this.setReference(pageNumber);
        }
      }
    },

    nextPage() {
      if (this.canNext) {
        const pageNumber = this.value + 1;

        this.selectPage(pageNumber);

        if (!this.pages.includes(pageNumber)) {
          this.setReference(pageNumber);
        }
      }
    },

    setReference(reference) {
      this.reference = reference;
    },
  },
};
</script>

<style lang="scss" scoped>
@import "../../assets/scss/unnnic.scss";

.pagination {
  display: flex;
  align-items: center;
  column-gap: $unnnic-spacing-inline-xs;
  user-select: none;
}
</style>
