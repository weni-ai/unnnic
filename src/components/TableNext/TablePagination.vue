<template>
  <section
    class="table-pagination"
  >
    <p
      class="table-pagination__count"
    >
      {{ tablePagination }}
    </p>

    <Pagination
      :value="value"
      @input="$emit('input', $event)"
      :max="pages"
      :show="5"
    />
  </section>
</template>

<script>
import Pagination from '../Pagination/Pagination.vue';
import UnnnicI18n from '../../mixins/i18n';

export default {
  name: 'TablePagination',

  mixins: [UnnnicI18n],

  components: {
    Pagination,
  },

  props: {
    value: {
      type: Number,
      required: true,
    },
    total: {
      type: Number,
      required: false,
    },
    interval: {
      type: Number,
      required: true,
    },
  },

  data() {
    return {
      defaultTranslations: {
        pagination: {
          'pt-br': '{from} - {to} de {total}',
          en: '{from} - {to} of {total}',
          es: '{from} - {to} de {total}',
        },
      },
    };
  },

  computed: {
    tablePagination() {
      const { value, interval, total } = this;

      return this.i18n('pagination', {
        from: total === 0 ? 0 : (value - 1) * interval + 1,
        to: Math.min(value * interval, total),
        total,
      });
    },

    pages() {
      return Math.ceil(this.total / this.interval);
    },
    showTotal() {
      return typeof this.total === 'number';
    },
  },
};
</script>

<style lang="scss" scoped>
@import "../../assets/scss/unnnic.scss";

.table-pagination {
  display: flex;
  align-items: center;
  justify-content: space-between;

  padding: $unnnic-spacing-sm $unnnic-spacing-nano;

  height: min-content;

  &__count {
    color: $unnnic-color-neutral-dark;
    font-size: $unnnic-font-size-body-gt;
    font-family: $unnnic-font-family-secondary;
  }
}
</style>
