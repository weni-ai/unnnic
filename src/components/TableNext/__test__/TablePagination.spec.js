import { mount } from '@vue/test-utils';
import { describe, it, expect, beforeEach } from 'vitest';
import TablePagination from '../TablePagination.vue';
import Pagination from '@/components/Pagination/Pagination.vue';

describe('TablePagination.vue', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(TablePagination, {
      props: {
        modelValue: 1,
        interval: 10,
        total: 100,
      },
    });
  });

  const findCountElement = () => wrapper.find('[data-testid="count"]');

  it.each([
    { modelValue: 2, total: 100, expectedText: '11 - 20 of 100' },
    { modelValue: 10, total: 95, expectedText: '91 - 95 of 95' },
    { modelValue: 1, total: 95, expectedText: '1 - 10 of 95' },
    { modelValue: 1, total: 0, expectedText: '0 - 0 of 0' },
  ])(
    'renders correct pagination text for modelValue: $modelValue and total: $total',
    async ({ modelValue, total, expectedText }) => {
      await wrapper.setProps({ modelValue, total });
      expect(findCountElement().text()).toContain(expectedText);
    },
  );

  it('calculates pages correctly', async () => {
    await wrapper.setProps({ total: 55, interval: 10 });
    const pagination = wrapper.vm.pages;
    expect(pagination).toBe(6); // 55 items / 10 per page = 6 pages
  });

  it('renders correct pagination text for the last page', async () => {
    await wrapper.setProps({ modelValue: 10, total: 95, interval: 10 });
    expect(findCountElement().text()).toContain('91 - 95 of 95');
  });

  it('renders correct pagination text for the first page', async () => {
    await wrapper.setProps({ modelValue: 1, total: 95, interval: 10 });
    expect(findCountElement().text()).toContain('1 - 10 of 95');
  });

  it('emits update:model-value event when pagination changes', async () => {
    await wrapper.findComponent(Pagination).vm.$emit('update:model-value', 2);

    expect(wrapper.emitted()['update:model-value']).toBeTruthy();
    expect(wrapper.emitted()['update:model-value'][0]).toEqual([2]);
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
