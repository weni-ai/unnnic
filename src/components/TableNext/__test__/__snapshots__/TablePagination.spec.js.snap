// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TablePagination.vue > matches the snapshot 1`] = `
"<section data-v-e29458bb="" class="table-pagination">
  <p data-v-e29458bb="" class="table-pagination__count" data-testid="count">1 - 10 of 100</p>
  <div data-v-0eeff8cb="" data-v-e29458bb="" class="pagination"><button data-v-ceff2a60="" data-v-0eeff8cb="" data-test="previous-button" disabled="" class="unnnic-button unnnic-button--size-small unnnic-button--tertiary unnnic-button--icon-on-center">
      <!--v-if-->
      <!--v-if--><span data-v-26446d8e="" data-v-ceff2a60="" class="unnnic-icon material-symbols-rounded unnnic-icon-scheme--neutral-clean unnnic-icon-size--sm unnnic-icon__size--sm" data-testid="icon-center" translate="no" style="visibility: visible;">chevron_left</span><span data-v-ceff2a60="" class="unnnic-button__label" style="visibility: visible;" data-testid="button-label"> </span>
      <!--v-if-->
    </button><button data-v-ceff2a60="" data-v-0eeff8cb="" data-test="page-button" class="unnnic-button unnnic-button--size-small unnnic-button--secondary">
      <!--v-if-->
      <!--v-if-->
      <!--v-if--><span data-v-ceff2a60="" class="unnnic-button__label" style="visibility: visible;" data-testid="button-label"> 1</span>
      <!--v-if-->
    </button><button data-v-ceff2a60="" data-v-0eeff8cb="" data-test="page-button" class="unnnic-button unnnic-button--size-small unnnic-button--tertiary">
      <!--v-if-->
      <!--v-if-->
      <!--v-if--><span data-v-ceff2a60="" class="unnnic-button__label" style="visibility: visible;" data-testid="button-label"> 2</span>
      <!--v-if-->
    </button><button data-v-ceff2a60="" data-v-0eeff8cb="" data-test="page-button" class="unnnic-button unnnic-button--size-small unnnic-button--tertiary">
      <!--v-if-->
      <!--v-if-->
      <!--v-if--><span data-v-ceff2a60="" class="unnnic-button__label" style="visibility: visible;" data-testid="button-label"> 3</span>
      <!--v-if-->
    </button><button data-v-ceff2a60="" data-v-0eeff8cb="" data-test="page-button" class="unnnic-button unnnic-button--size-small unnnic-button--tertiary">
      <!--v-if-->
      <!--v-if-->
      <!--v-if--><span data-v-ceff2a60="" class="unnnic-button__label" style="visibility: visible;" data-testid="button-label"> ...</span>
      <!--v-if-->
    </button><button data-v-ceff2a60="" data-v-0eeff8cb="" data-test="page-button" class="unnnic-button unnnic-button--size-small unnnic-button--tertiary">
      <!--v-if-->
      <!--v-if-->
      <!--v-if--><span data-v-ceff2a60="" class="unnnic-button__label" style="visibility: visible;" data-testid="button-label"> 10</span>
      <!--v-if-->
    </button><button data-v-ceff2a60="" data-v-0eeff8cb="" data-test="next-button" class="unnnic-button unnnic-button--size-small unnnic-button--tertiary unnnic-button--icon-on-center">
      <!--v-if-->
      <!--v-if--><span data-v-26446d8e="" data-v-ceff2a60="" class="unnnic-icon material-symbols-rounded unnnic-icon-scheme--neutral-dark unnnic-icon-size--sm unnnic-icon__size--sm" data-testid="icon-center" translate="no" style="visibility: visible;">chevron_right</span><span data-v-ceff2a60="" class="unnnic-button__label" style="visibility: visible;" data-testid="button-label"> </span>
      <!--v-if-->
    </button></div>
</section>"
`;
